README.md
# opx-sai-api
This repository contains all public API declarations for the switch abstraction interface (SAI) API. The network abstraction service (NAS) component utilizes this SAI API for network processor unit (NPU) related operations (such as port, L2, and L3 related). 

## Package
- opx-sai-api-dev\_*version*\_*arch*.deb — Header files for public API  

See [Architecture](https://github.com/open-switch/opx-docs/wiki/Architecture) for more information on the SAI API.  

(c) 2017 Dell


debian/changelog
opx-sai-api (38.1.0) unstable; urgency=medium

  * Feature: SAI API VXLAN overlay ecmp changes
  * Update: Scaled VXLAN mode support
  * Update: Introduced a new SAI attribute for polling S,G entry actively forwarded in H/W or not

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 18 Mar 2019 11:45:33 -0800

opx-sai-api (37.0.0) unstable; urgency=medium

  * Feature: Buffer Statistics

 -- Dell EMC <ops-dev@lists.openswitch.net>  Thu, 15 Nov 2018 11:45:33 -0800

opx-sai-api (36.0.0) unstable; urgency=medium

  * Feature: Static L2 VxLAN Support
  * Feature: SAI 1.2
  * Update: Add IPv6 Extended prefix(>/64) route entry size
  * Update: SAI-API Attributes for supporting split horizon
  * Update: Additional IP stat attributes
  * Update: Display TCAM utilization
  * Update: Add support for match brige type in ACL
  * Update: Add egress filter attribute to bridgeport
  * Bugfix: Fix the VLAN Id attribute

 -- Dell EMC <ops-dev@lists.openswitch.net>  Tue, 06 Nov 2018 11:45:33 -0800

opx-sai-api (35.1.0+opx2) unstable; urgency=medium

  * Update: Support for IPv6 extended prefix (>64) route entry

 -- Dell EMC <ops-dev@lists.openswitch.net>  Fri, 10 Aug 2018 11:45:33 -0800

opx-sai-api (35.1.0+opx1) unstable; urgency=medium

  * Update: IP Stat support
  * Bugfix: Unit Test fixes

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 9 Jul 2018 11:45:33 -0800

opx-sai-api (35.1.0) unstable; urgency=medium

  * Update: Upgrade to SAI 1.2
  * Update: Support for OCP SAI verification tool
  * Update: SAI metadata generation tool from OCP
  * Feature: Global ECN Support for 4200

 -- Dell EMC <ops-dev@lists.openswitch.net>  Wed, 23 May 2018 11:45:33 -0800

opx-sai-api (32.0.0+opx1) unstable; urgency=medium

  * Update: Add compiler/linker hardening flags

 -- Dell EMC <ops-dev@lists.openswitch.net>  Thu, 17 May 2018 11:45:33 -0800

opx-sai-api (32.0.0) unstable; urgency=medium

  * Update: Added vendor extension to get PG/Q mapping to XPE.
  * Update: Added HDRM pool buffer counters.
  * Update: Added vxlan unit tests
  * Bugfix: Miscellaneous bug fixes

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 12 Dec 2017 11:45:33 -0800

opx-sai-api (31.0.0) unstable; urgency=medium

  * Update: Synced with latest saitunnel.h and saitypes.h
  * Update: Changes to support WRED at Queue, Port and Buffer Pool level
  * Update: SAI API for WRED - SAI v0.9.6
  * Update: switch attribute
  * Update: Added new attribute to saifcswitch.
  * Bugfix: Miscellaneous bug fixes
  * Cleanup: Miscellaneous cleanup

 -- Dell EMC <ops-dev@lists.openswitch.net>  Mon, 2 Oct 2017 11:45:33 -0800

opx-sai-api (26.1.0) unstable; urgency=medium

  * Update: SAI 1.0 changes
  * Update: SAI 0.9.6 changes
  * Feature: SAI FC Module API support
  * Bugfix: Miscellaneous Bugfixes
  * Cleanup: Miscellaneous Cleanup

 -- Dell EMC <ops-dev@lists.openswitch.net>  Fri, 16 Jun 2017 11:45:33 -0800


opx-sai-api (1.0.1) unstable; urgency=medium

  * Initial release.

 -- Dell EMC <ops-dev@lists.openswitch.net>  Wed, 01 Feb 2017 09:45:33 -0800


inc/sai.h
/**
 * Copyright (c) 2014 Microsoft Open Technologies, Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License"); you may
 *    not use this file except in compliance with the License. You may obtain
 *    a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 *    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR
 *    CONDITIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT
 *    LIMITATION ANY IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS
 *    FOR A PARTICULAR PURPOSE, MERCHANTABILITY OR NON-INFRINGEMENT.
 *
 *    See the Apache Version 2.0 License for specific language governing
 *    permissions and limitations under the License.
 *
 *    Microsoft would like to thank the following companies for their review and
 *    assistance with these files: Intel Corporation, Mellanox Technologies Ltd,
 *    Dell EMC, Facebook, Inc., Marvell International Ltd.
 *
 * @file    sai.h
 *
 * @brief   This module defines an entry point into Switch Abstraction Interface (SAI)
 */

#if !defined (__SAI_H_)
#define __SAI_H_

#include "saiacl.h"
#include "saibridge.h"
#include "saibuffer.h"
#include "saifdb.h"
#include "saihash.h"
#include "saihostif.h"
#include "saiipmcgroup.h"
#include "saiipmc.h"
#include "sail2mcgroup.h"
#include "sail2mc.h"
#include "sailag.h"
#include "saimcastfdb.h"
#include "saimirror.h"
#include "saimpls.h"
#include "saineighbor.h"
#include "sainexthopgroup.h"
#include "sainexthop.h"
#include "saiobject.h"
#include "saipolicer.h"
#include "saiport.h"
#include "saiqosmap.h"
#include "saiqueue.h"
#include "sairoute.h"
#include "sairouterinterface.h"
#include "sairpfgroup.h"
#include "saisamplepacket.h"
#include "saischedulergroup.h"
#include "saischeduler.h"
#include "saisegmentroute.h"
#include "saistatus.h"
#include "saistp.h"
#include "saiswitch.h"
#include "saitam.h"
#include "saitunnel.h"
#include "saitypes.h"
#include "saiudf.h"
#include "saivirtualrouter.h"
#include "saivlan.h"
#include "saiwred.h"
#include "saiuburst.h"

/**
 * @defgroup SAI SAI - Entry point specific API definitions.
 *
 * @{
 */

/**
 * @brief Defined API sets have assigned IDs.
 * @flags Contains flags
 *
 * If specific API method table changes in any way (method signature, number of
 * methods), a new ID needs to be created (e.g. VLAN2) and old API still may
 * need to be supported for compatibility with older adapter hosts.
 */
typedef enum _sai_api_t
{
    SAI_API_UNSPECIFIED      =  0, /**< unspecified API */
    SAI_API_SWITCH           =  1, /**< sai_switch_api_t */
    SAI_API_PORT             =  2, /**< sai_port_api_t */
    SAI_API_FDB              =  3, /**< sai_fdb_api_t */
    SAI_API_VLAN             =  4, /**< sai_vlan_api_t */
    SAI_API_VIRTUAL_ROUTER   =  5, /**< sai_virtual_router_api_t */
    SAI_API_ROUTE            =  6, /**< sai_route_api_t */
    SAI_API_NEXT_HOP         =  7, /**< sai_next_hop_api_t */
    SAI_API_NEXT_HOP_GROUP   =  8, /**< sai_next_hop_group_api_t */
    SAI_API_ROUTER_INTERFACE =  9, /**< sai_router_interface_api_t */
    SAI_API_NEIGHBOR         = 10, /**< sai_neighbor_api_t */
    SAI_API_ACL              = 11, /**< sai_acl_api_t */
    SAI_API_HOSTIF           = 12, /**< sai_hostif_api_t */
    SAI_API_MIRROR           = 13, /**< sai_mirror_api_t */
    SAI_API_SAMPLEPACKET     = 14, /**< sai_samplepacket_api_t */
    SAI_API_STP              = 15, /**< sai_stp_api_t */
    SAI_API_LAG              = 16, /**< sai_lag_api_t */
    SAI_API_POLICER          = 17, /**< sai_policer_api_t */
    SAI_API_WRED             = 18, /**< sai_wred_api_t */
    SAI_API_QOS_MAP          = 19, /**< sai_qos_map_api_t */
    SAI_API_QUEUE            = 20, /**< sai_queue_api_t */
    SAI_API_SCHEDULER        = 21, /**< sai_scheduler_api_t */
    SAI_API_SCHEDULER_GROUP  = 22, /**< sai_scheduler_group_api_t */
    SAI_API_BUFFER           = 23, /**< sai_buffer_api_t */
    SAI_API_HASH             = 24, /**< sai_hash_api_t */
    SAI_API_UDF              = 25, /**< sai_udf_api_t */
    SAI_API_TUNNEL           = 26, /**< sai_tunnel_api_t */
    SAI_API_L2MC             = 27, /**< sai_l2mc_api_t */
    SAI_API_IPMC             = 28, /**< sai_ipmc_api_t */
    SAI_API_RPF_GROUP        = 29, /**< sai_rpf_group_api_t */
    SAI_API_L2MC_GROUP       = 30, /**< sai_l2mc_group_api_t */
    SAI_API_IPMC_GROUP       = 31, /**< sai_ipmc_group_api_t */
    SAI_API_MCAST_FDB        = 32, /**< sai_mcast_fdb_api_t */
    SAI_API_BRIDGE           = 33, /**< sai_bridge_api_t */
    SAI_API_TAM              = 34, /**< sai_tam_api_t */
    SAI_API_SEGMENTROUTE     = 35, /**< sai_segmentroute_api_t */
    SAI_API_MPLS             = 36, /**< sai_mpls_api_t */
    SAI_API_UBURST           = 37, /**< sai_uburst_api_t */

    /** @ignore */
    SAI_API_MAX              = 38, /**< total number of APIs */
    SAI_API_CUSTOM_RANGE_START = 0x1000,
    /** Extension module provided by Dell Inc. SAI implementation */
    SAI_API_FC_SWITCH = SAI_API_CUSTOM_RANGE_START,
    SAI_API_FC_PORT,
    SAI_API_CUSTOM_RANGE_END,
} sai_api_t;

/**
 * @brief Defines log level
 */
typedef enum _sai_log_level_t
{
    /** Log Level Debug */
    SAI_LOG_LEVEL_DEBUG            = 0,

    /** Log Level Info */
    SAI_LOG_LEVEL_INFO             = 1,

    /** Log Level Notice */
    SAI_LOG_LEVEL_NOTICE           = 2,

    /** Log level Warning */
    SAI_LOG_LEVEL_WARN             = 3,

    /** Log Level Error */
    SAI_LOG_LEVEL_ERROR            = 4,

    /** Log Level Critical */
    SAI_LOG_LEVEL_CRITICAL         = 5

} sai_log_level_t;

typedef const char* (*sai_profile_get_value_fn)(
        _In_ sai_switch_profile_id_t profile_id,
        _In_ const char *variable);

typedef int (*sai_profile_get_next_value_fn)(
        _In_ sai_switch_profile_id_t profile_id,
        _Out_ const char** variable,
        _Out_ const char** value);

/**
 * @brief Method table that contains function pointers for services exposed by
 * adapter host for adapter.
 */
typedef struct _sai_service_method_table_t
{
    /**
     * @brief Get variable value given its name
     */
    sai_profile_get_value_fn        profile_get_value;

    /**
     * @brief Enumerate all the K/V pairs in a profile.
     *
     * Pointer to NULL passed as variable restarts enumeration. Function
     * returns 0 if next value exists, -1 at the end of the list.
     */
    sai_profile_get_next_value_fn   profile_get_next_value;

} sai_service_method_table_t;

/**
 * @brief Adapter module initialization call
 *
 * This is NOT for SDK initialization.
 *
 * @param[in] flags Reserved for future use, must be zero
 * @param[in] services Methods table with services provided by adapter host
 *
 * @return #SAI_STATUS_SUCCESS on success, failure status code on error
 */
sai_status_t sai_api_initialize(
        _In_ uint64_t flags,
        _In_ const sai_service_method_table_t *services);

/**
 * @brief Retrieve a pointer to the C-style method table for desired SAI
 * functionality as specified by the given sai_api_id.
 *
 * @param[in] sai_api_id SAI API ID
 * @param[out] api_method_table Caller allocated method table. The table must
 * remain valid until the sai_api_uninitialize() is called.
 *
 * @return #SAI_STATUS_SUCCESS on success, failure status code on error
 */
sai_status_t sai_api_query(
        _In_ sai_api_t sai_api_id,
        _Out_ void** api_method_table);

/**
 * @brief Uninitialize adapter module. SAI functionalities,
 * retrieved via sai_api_query() cannot be used after this call.
 *
 * @return #SAI_STATUS_SUCCESS on success, failure status code on error
 */
sai_status_t sai_api_uninitialize(void);

/**
 * @brief Set log level for SAI API module
 *
 * The default log level is #SAI_LOG_LEVEL_WARN.
 *
 * @param[in] sai_api_id SAI API ID
 * @param[in] log_level Log level
 *
 * @return #SAI_STATUS_SUCCESS on success, failure status code on error
 */
sai_status_t sai_log_set(
        _In_ sai_api_t sai_api_id,
        _In_ sai_log_level_t log_level);

/**
 * @brief Query SAI object type.
 *
 * @param[in] sai_object_id Object id
 *
 * @return #SAI_OBJECT_TYPE_NULL when sai_object_id is not valid.
 * Otherwise, return a valid SAI object type SAI_OBJECT_TYPE_XXX.
 */
sai_object_type_t sai_object_type_query(
        _In_ sai_object_id_t sai_object_id);

/**
 * @brief Query SAI switch id.
 *
 * @param[in] sai_object_id Object id
 *
 * @return #SAI_NULL_OBJECT_ID when sai_object_id is not valid.
 * Otherwise, return a valid SAI_OBJECT_TYPE_SWITCH object on which
 * provided object id belongs. If valid switch id object is provided
 * as input parameter it should return itself.
 */
sai_object_id_t sai_switch_id_query(
        _In_ sai_object_id_t sai_object_id);

/**
 * @brief Generate dump file. The dump file may include SAI state information and vendor SDK information.
 *
 * @param[in] dump_file_name Full path for dump file
 *
 * @return #SAI_STATUS_SUCCESS on success, failure status code on error
 */
sai_status_t sai_dbg_generate_dump(
        _In_ const char *dump_file_name);

/**
 * @}
 */
#endif /** __SAI_H_ */
